src/librustc/ty/context.rs

```diff
@@ -1602,6 +1602,26 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         &self,
         scope_def_id: DefId,
     ) -> Option<Ty<'tcx>> {
+        // HACK: `type_of_def_id()` will fail on these (#55796), so return None
+        let node_id = self.hir.as_local_node_id(scope_def_id).unwrap();
+        match self.hir.get(node_id) {
+            Node::Item(item) => {
+                match item.node {
+                    ItemKind::Trait(..)
+                    | ItemKind::TraitAlias(..)
+                    | ItemKind::Mod(..)
+                    | ItemKind::ForeignMod(..)
+                    | ItemKind::GlobalAsm(..)
+                    | ItemKind::ExternCrate(..)
+                    | ItemKind::Use(..) => {
```

Comment:
Realistically, the only check _actually_ needed here is `Trait`, but I want to be conservative in behavior, just skipping instead of only making the ICE harder to trigger.

Reply:
We could also swap the logic and return `None` on everything that isn't `ItemKind::Fn`.

Reply:
@oli-obk done.